### DISCUSSION TITLE
User-consumability

### DISCUSSION THREAD

**Discussion by tjjfvi** [typed-ember/semver#3] [General]
Consider the imaginary package `frobnicate`, maintained by a friendly guy named Fred. It's a simple package, on the surface. Fred has chosen to follow Semantic Versioning for TypeScript Types in `frobnicate`. This is the source code of `frobnicate@1.0.0`:

```ts
export type FrobnicateOptions = {
  intensity: number,
  rng: () => number,
}

export function frobnicate(options: FrobnicateOptions) {
  // ...
}
```

This is perfectly reasonable code. The most notable thing about it is that it exports `FrobnicateOptions`, but this is usually considered good practice, as it allows users to use it in cases such as:
```ts
const frobnicateOptions: FrobnicateOptions = {
  intensity: 5,
  rng: () => Math.random() ** 2,
}
```

If `FrobnicateOptions` is not exported, users must resort to something like `Parameters<typeof frobnicate>[0]`, which is rather unattractive.

Soon, Fred gets a feature request – in the `rng` function, a user wants to be able to know some degree of state, without having to track it manually. This is a simple fix, Fred realizes; he can simply pass a number indicating how many times the function has been called before in this frobnication pass.

```ts
export type FrobnicateOptions = {
  intensity: number,
  rng: (callNumber: number) => number,
}

export function frobnicate(options: FrobnicateOptions) {
  // ...
}
```

Now, someone can create an `rng` such as `i => Math.random() * i`.

Fred is happy with this change, and seeks to release it as `frobnicate@1.1.0`.

However, to his horror, he realizes that **this would be a breaking change**:

> A change to any object type (user constructible or not) is breaking when:
> - a non-`readonly` object property's type changes in any way

[Source](https://www.semver-ts.org/#:~:text=A%20change%20to%20any,changes%20in%20any%20way)

---

This would usually not be considered a breaking change, as `() => number` is assignable to `(x: number) => number`, and thus no type errors would occur in constructing code.

However, the current proposal treats all types as *user-consumable*, and so this change would break user code such as:
```ts
const options: FrobnicateOptions = {
  intensity: 5,
  rng: () => Math.random() ** 2,
}

options.rng()
```

Because now the type of `options.rng` expects a parameter.

---

I propose the following additions to the spec:

A new concept, *user-consumability*.

Exported values from a library are, by default, user-consumable but not user-constructible, unless specifically noted otherwise in their type.

User-consumability of a type entails the following permissions:
- Reading properties of a value of the type
- Writing properties of a value of the type, if they are not `readonly`
- Calling a value of the type, if it is a function type

The terms *user-constructible* and *user-consumable* refer to ways of constructing and consuming the types without going through the library; for example, user-constructible types allow object literal construction, and user-consumable types allow property access.

In particular, "constructing" and "consuming" values by way of user-consumable library functions is always allowed.

Specifically:
- Users can "consume" a non-user-consumable type by passing it to a user-consumable function exported from the library
- Users can "construct" a non-user-constructible type by getting it from a user-consumable function exported from the library
- A user-consumable function may take non-user-constructible arguments, and its return type may be non-user-consumable.
- A user-constructible function may take non-user-consumable arguments, and its return type may be non-user-constructible.

**Comment by chriskrycho** [comment]
Thanks for writing this up! I have two overarching notes here as I’ve been mulling on it since you filed it yesterday, so I'm going to split those into two separate posts here for discussion!

---

I *think* in this case this falls under the user-constructibility notion: the interface is exported explicitly for the sake of users making one. (I’ll come back to this in the next reply!)

“Consumability” is implicitly in the text as it stands—it is another way of naming what the spec describes as “public” types. Accordingly, it’s not *totally* clear to me that it’s something we need to add more terminology. My own read is that if you want a type not to be “user-consmable” you should handle that using privacy (whether via `private`, `#private`, sneaky use of `WeakMap`s, using opaque types in your public API, etc.). But I could be wrong! I have a “curse of knowledge” problem here in that I have lived with this for a long time. Making it explicit might still be helpful.

I’d also be curious to hear feedback from @dfreeman and @jamescdavis here.

**Comment by chriskrycho** [comment]
## Widening functions

The language here is, as your example shows, overly-narrow. (It’s just plain wrong!) There *is* a way to make the change here *without* its being breaking, even for a user-constructible type, but we need to frame it correctly. Here, as you note, it’s still a perfectly-valid extension, under both the “accepts—more” intuition and what the type system actually accepts. (For others following along, [here’s a playground][1] demonstrating as much!)

I originally had something here much closer to what you suggested, and had made the stronger statement currently in the text because I hadn’t come up with a counterexample, but this one matches, *as does* widening that type to a union:

```ts
export type FrobnicateOptions = {
  intensity: number;
  rng:
    | (() => number)
    | ((callNumber: number) => number) 
    | ((options: unknown) => number)
    | ((prev: number, seed?: number) => number);
};

// ...
```

The spec text as it stands does not properly consider functions specifically in this case. But widening for non-function types in these cases *is* breaking! So the text needs to handle functions distinctly from other properties, both for user-constructible types and for user-consumable types (adopting your nomenclature for now).


[1]: https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEA7KBbEBnADlMCDyMAlgOaHITwDeAUPPCAB6YD2MALvGwJ6YIBiMZgCNEhMFDYg8mNoWaJ08ALxVadeGUkLC3AFxIAriiEgYAbjV0YiYvoAUASmUA+Q8dMW6AXwtrGLdngAMwNEMFl5YMERMQkQO2YZOQV9AWFRcUlpCIUHfQA3ZkJgCy9qalBIWARkNCwcBABVTGA44FU6f1YObl54NJjMqST5RRUadQ1ELXQdLn1EIxNzS3hrW3g7cQgIADkl0wWDmCclV0X3Fe9fTqZu4NDw5Kj02MkEkZT+6Iy47OT0Hl4IViqVygB6cHwACS0xA02SUB2XCQzA4pHy8PgUE4PAQUEQiDREmSFjAow4iRyYw6Uxmc30AFYADRqdb2U6uACyEgAFgA6GAE4DMFCOeAAKgl8AATKyytQqQDBTZHBCoVB0LNiMghBAEGxmPBWCQyEjmdjMJhBNbCHEICjttA9SBqPrKUQNgRTeR+QNfllPjSlaMLN7SL6gj83vEQ7lfJDsVqSLr9ZwjQYWm03SAOJnWpJgPpmgWQMA-dGhv9RspjUGLCW2vyo68hh9qQ4LEA

[2]: https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEA7KBbEBnADlMCDyMAlgOaHITwDeAUPPCAB6YD2MALvGwJ6YIBiMZgCNEhMFDYg8mNoWaJ08ALxVadeGUkLC3AFxIAriiEgYAbjV0YiYvoAUASmUA+Q8dMW6AXwtrGLdngAMwNEMFl5YMERMQkQO2YZOQV9AWFRcUlpCIUHfQA3ZkJgCy9qalBIWARkNCwcBABVTGA44FU6f1YObl54NJjMqST5RRUadQ1ELXQdLn1EIxNzS3hrW1W6AB94OztxCAgAOSXTBdOYJyVXRfdL+E34Hb3EnPR9UIBrRGYAd0QrjcLg5Hs87Jg4PlzncADTwdAgEDAAD80OWgLc6M88B85U6TG6wVC4WSUXSsUkCRGKX60QycWyyXQeXghWKpXKAHpOfAAJLTEDTZJQQ5cJDMDikfKC+BQTg8BBQRA-NgSZIWMCjDivJnKDpTGZzfQAVhhanW9gxAFkJAALAB0MCVwGYKEc8AAVB74AAmM1lag60aOmyOLk8qDoWbEZBCCAINjMeCsEhkEVwqCYCGJIhxCBig7QOMgajx7VEWzwAip8j2gb0rLUsbJpsWaukWtBOkU+JB3K+bmyqMkWPxzhJgwtNqlkAcSetSTAfTNBdIuvdoaM0Z6vvoCwrtr2rvkoZUt4OCxAA

**Comment by tjjfvi** [comment]
Sorry, I completely forgot about this; I'll try to take a look in the next week or so and get back to you on this.

**Reply by chriskrycho** [reply]
No worries! We’re all busy, totally understand. There’s a reason I only came back to all of this today myself.

**Comment by tjjfvi** [comment]
> I originally had something here much closer to what you suggested, and had made the stronger statement currently in the text because I hadn’t come up with a counterexample, but this one matches, as does widening that type to a union:

Given an original type of
```ts
export type FrobnicateOptions = {
  intensity: number,
  rng: () => number,
}
```
in the text as it stands, both
```ts
export type FrobnicateOptions = {
  intensity: number,
  rng: (callNumber: number) => number,
}
```
and
```ts
export type FrobnicateOptions = {
  intensity: number;
  rng:
    | (() => number)
    | ((callNumber: number) => number) 
    | ((options: unknown) => number)
    | ((prev: number, seed?: number) => number);
};
```
are breaking changes, as they both break user code such as
```ts
function foo(options: FrobnicateOptions){
  options.rng()
}
```

However, if `FrobnicateOptions` was non-user-consumable, the above user code would not be valid, and so both of our changes would be non-breaking.
